test/cloud/yaml_test.rb

```diff
@@ -0,0 +1,82 @@
+require File.expand_path(File.join(__FILE__, '../..', 'test_helper'))
+require 'rubber/cloud/yaml'
+
+class YAMLTest < Test::Unit::TestCase
+  context 'yaml' do
+    setup do
+      ENV["YAML_DATABASE"] = "/tmp/#{SecureRandom.uuid}"
+
+      env = Rubber::Configuration::Environment::BoundEnv.new({}, nil, nil, nil)
+      @cloud = Rubber::Cloud::YAML.new(env, nil)
+    end
+
+    teardown do
+      FileUtils.rm_f(ENV.delete("YAML_DATABASE"))
+    end
+
+    should 'create instance' do
+      db = [
+        Rubber::Cloud::YAML::Instance.new(SecureRandom.uuid, Rubber::Cloud::YAML::AVAILABLE, 'dc0', '127.0.0.1', '10.0.0.1', nil, nil),
+      ]
+
+      Rubber::Cloud::YAML.persist_database(db, ENV["YAML_DATABASE"])
+
+      assert_equal db.first.id, @cloud.create_instance('', '', '', '', '', '')
+      instance = @cloud.describe_instances(db.first.id).first
+
+      # ensure creating the instance populated the two missing fields.
+      assert_equal instance[:provider], 'YAML'
+      assert_equal instance[:platform], Rubber::Platforms::LINUX
+    end
+
+    context 'describe_instances' do
+      should 'be able to describe all instances if no instance id is provided' do
+        db = [
+          Rubber::Cloud::YAML::Instance.new(SecureRandom.uuid, Rubber::Cloud::YAML::ACTIVE, 'dc0', '127.0.0.1', '10.0.0.1', nil, nil),
+          Rubber::Cloud::YAML::Instance.new(SecureRandom.uuid, Rubber::Cloud::YAML::STOPPED, 'dc1', '127.0.0.2', '10.0.0.2', nil, nil),
+          Rubber::Cloud::YAML::Instance.new(SecureRandom.uuid, Rubber::Cloud::YAML::AVAILABLE, 'dc1', '127.0.0.3', '10.0.0.3', nil, nil),
+        ]
+
+        # Load DB.
+        Rubber::Cloud::YAML.persist_database(db, ENV["YAML_DATABASE"])
+
+        instances = @cloud.describe_instances
+        assert_equal 2, instances.count
```

Comment:
Yes, the db has 3, however only 2 are in a 'in use' state. describe_instances only returns instances you as a user of rubber have created, not what is available. Lets say I have 5 servers available for use. I consume one to create a database. Now I have 4 servers available and 1 in use. describe_instances only returns the ones in use.